### PR TITLE
[android] Make Drape the source of truth for the location mode changes

### LINKED ISSUE #1128: Do not show "Enable location services" on every launch
Some users do not enable Location services at all and always see the "Please enable Location Services" annoying dialog on every app launch. The dialog appears anyway if the user presses the "my location" button in the app, so for better UX it makes sense to show dialog only once and store this fact in the settings, but reset this setting again if the location was successfully acquired (e.g. user enabled location services).

The main idea is to show this warning only once, if the user does not enable location services, and show it again if services were enabled, and then disabled again.

### LINKED ISSUE #6098: Pop up to continue to search location is visible after I deny to search location
**Steps to reproduce**
1. Start OM (Need Google Play Services Enabled)
2. Deny Google Play pop-up
3. Wait 1 minute -> Pop-up to continue to search location is visible

**System information:**
 - Operating system and its version: Android 13
 - Organic Maps version: 2023.09.22-25-Web-debug
 - Device Model: Pixel 6

### PR DESCRIPTION
**Pull request by rtsisyk**
[android] Make Drape the source of truth for the location mode changes

Drape changes the location mode under the hood in multiple cases, which
creates the discrepancy between Android <-> Drape state. Trust Drape
and do all location mode transitions only via onMyPositionModeChanged().

Follow up #1128 #4376 #6098 #6465 #6466

Signed-off-by: Roman Tsisyk <roman@tsisyk.com>

### CODE REVIEW

**Review by Jean-BaptisteC — CHANGES_REQUESTED** [review]
❌ Pixel 6 - Android 14

- Start OM
- Use location button -> Google Play services dialog appears
- Rotate device -> Crash

```
E  vulkan/android_vulkan_context_factory.cpp:428 GetWidth(): CHECK(IsValid()) 
A  /home/runner/work/organicmaps/organicmaps/android/app/src/main/cpp/app/organicmaps/vulkan/android_vulkan_context_factory.cpp:428: int android::AndroidVulkanContextFactory::GetWidth() const: assertion "false" failed
A  Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 4074 (ganicmaps.debug), pid 4074 (ganicmaps.debug)
A  Cmdline: app.organicmaps.debug
A  pid: 4074, tid: 4074, name: ganicmaps.debug  >>> app.organicmaps.debug <<<
A        #02 pc 00000000020a31a4  /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/lib/arm64/liborganicmaps.so (android::AndroidVulkanContextFactory::GetWidth() const+252) (BuildId: 9779bba2ddb5b7f256a597a3097fddfd2774f86d)
A        #03 pc 00000000020a4020  /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/lib/arm64/liborganicmaps.so (android::Framework::Resize(_JNIEnv*, _jobject*, int, int)+92) (BuildId: 9779bba2ddb5b7f256a597a3097fddfd2774f86d)
A        #04 pc 00000000020b39a0  /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/lib/arm64/liborganicmaps.so (Java_app_organicmaps_Map_nativeSurfaceChanged+60) (BuildId: 9779bba2ddb5b7f256a597a3097fddfd2774f86d)
A        #10 pc 000000000000f728  [anon:dalvik-classes4.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk!classes4.dex] (app.organicmaps.Map.onSurfaceChanged+0)
A        #15 pc 000000000000f218  [anon:dalvik-classes4.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk!classes4.dex] (app.organicmaps.MapFragment.surfaceChanged+0)
A        #74 pc 0000000000253788  [anon:dalvik-classes.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk] (androidx.coordinatorlayout.widget.CoordinatorLayout.layoutChild+0)
A        #79 pc 000000000025419c  [anon:dalvik-classes.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk] (androidx.coordinatorlayout.widget.CoordinatorLayout.onLayoutChild+0)
A        #84 pc 000000000025411c  [anon:dalvik-classes.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk] (androidx.coordinatorlayout.widget.CoordinatorLayout.onLayout+0)
```

Another weird thing
- Start OM
- Accept Google Play Services pop-up
- Disable location
- Move to the map -> Google Play Services pop-up appears

**Comment by rtsisyk** [comment]
> ❌ Pixel 6 - Android 14
> 
>     * Start OM
> 
>     * Use location button -> Google Play services dialog appears
> 
>     * Rotate device -> Crash
> 
> 
> ```
> E  vulkan/android_vulkan_context_factory.cpp:428 GetWidth(): CHECK(IsValid()) 
> A  /home/runner/work/organicmaps/organicmaps/android/app/src/main/cpp/app/organicmaps/vulkan/android_vulkan_context_factory.cpp:428: int android::AndroidVulkanContextFactory::GetWidth() const: assertion "false" failed
> A  Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 4074 (ganicmaps.debug), pid 4074 (ganicmaps.debug)
> A  Cmdline: app.organicmaps.debug
> A  pid: 4074, tid: 4074, name: ganicmaps.debug  >>> app.organicmaps.debug <<<
> A        #02 pc 00000000020a31a4  /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/lib/arm64/liborganicmaps.so (android::AndroidVulkanContextFactory::GetWidth() const+252) (BuildId: 9779bba2ddb5b7f256a597a3097fddfd2774f86d)
> A        #03 pc 00000000020a4020  /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/lib/arm64/liborganicmaps.so (android::Framework::Resize(_JNIEnv*, _jobject*, int, int)+92) (BuildId: 9779bba2ddb5b7f256a597a3097fddfd2774f86d)
> A        #04 pc 00000000020b39a0  /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/lib/arm64/liborganicmaps.so (Java_app_organicmaps_Map_nativeSurfaceChanged+60) (BuildId: 9779bba2ddb5b7f256a597a3097fddfd2774f86d)
> A        #10 pc 000000000000f728  [anon:dalvik-classes4.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk!classes4.dex] (app.organicmaps.Map.onSurfaceChanged+0)
> A        #15 pc 000000000000f218  [anon:dalvik-classes4.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk!classes4.dex] (app.organicmaps.MapFragment.surfaceChanged+0)
> A        #74 pc 0000000000253788  [anon:dalvik-classes.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk] (androidx.coordinatorlayout.widget.CoordinatorLayout.layoutChild+0)
> A        #79 pc 000000000025419c  [anon:dalvik-classes.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk] (androidx.coordinatorlayout.widget.CoordinatorLayout.onLayoutChild+0)
> A        #84 pc 000000000025411c  [anon:dalvik-classes.dex extracted in memory from /data/app/~~Dtmiw6hpPByBS4NUK3WpBw==/app.organicmaps.debug-xMGxbuOUPXCP0zF16tj3FQ==/base.apk] (androidx.coordinatorlayout.widget.CoordinatorLayout.onLayout+0)
> ```
> 
> Another weird thing
> 
>     * Start OM
> 
>     * Accept Google Play Services pop-up
> 
>     * Disable location
> 
>     * Move to the map -> Google Play Services pop-up appears

Are you sure that this issue happens because of the patch? I don't really understand the correlation... I will try to repro locally.

**Comment by Jean-BaptisteC** [comment]
Yes you're right, I'm able to reproduce on master, not on latest release.

File: android/app/src/main/java/app/organicmaps/location/LocationState.java [hunk 1]
```diff
@@ -63,17 +65,12 @@ static native void nativeLocationUpdated(long time, double lat, double lon, floa
...
-  }
-
-  static boolean hasLocation(int mode)
+  @Value
+  public static int getMode()
   {
-    return (mode > NOT_FOLLOW_NO_POSITION);
+    if (!Map.isEngineCreated())
+      throw new IllegalStateException("Location mode is undefined until engine is created");
+    return nativeGetMode();
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/location/LocationState.java:73]: No, it isn't. In this case any existing or new code that use the function will not work properly. This is the exact reason why we have this bug. autostartLocation() called from onResume() checked nativeGetMode() which returned  bougs mode = 0 (PENDING_POSITION) initially. Further initialization of drape changes the value to NO_FOLLOW_NO_LOCATION... The complain was that people didn't want to see location dialogs if location is "disabled"... We implement "location is disabled" via NO_FOLLOW_NO_POSITION... this is not right...

**Review by vng — COMMENTED** [review]
Discussed to remove PendingPosition -> NotFollowNoPosition.
Probably, it will solve the issue with ambiguity of NotFollowNoPosition state.

**Comment by rtsisyk** [comment]
> Discussed to remove PendingPosition -> NotFollowNoPosition transition. Probably, it will solve the issue with the ambiguity of NotFollowNoPosition state.

This part has been extracted to https://github.com/organicmaps/organicmaps/pull/6465.

File: drape_frontend/my_position_controller.cpp [hunk 2]
```diff
@@ -922,13 +922,13 @@ void MyPositionController::DeactivateRouting()
 
 void MyPositionController::CheckIsWaitingForLocation()
 {
-  if (IsWaitingForLocation() || m_mode == location::NotFollowNoPosition)
+  if (m_listener && IsWaitingForLocation())
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply drape_frontend/my_position_controller.cpp:925]: This part has been extracted to https://github.com/organicmaps/organicmaps/pull/6465.

**Review by Jean-BaptisteC — APPROVED** [review]
✅ Pixel 6 - Android 14
- Tested with Google Play Services

**Comment by rtsisyk** [comment]
The PR has been reworked to use Drape as the source of truth of all location mode changes. Please review and test thoroughly.

**Comment by rtsisyk** [comment]
Fixed false-positive lint error.

**Review by Jean-BaptisteC — APPROVED** [review]
✅ Pixel 6 - Android 14
- With Google Play Services
- Without Google Play Services

I have found a issue not related with this PR #6476

**Comment by rtsisyk** [comment]
Rebased.

File: android/app/src/main/java/app/organicmaps/location/LocationState.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (biodranik) [reply android/app/src/main/java/app/organicmaps/location/LocationState.java:73]: Can adding LOCATION_IS_DISABLED location mode solve this issue?

**Comment by rtsisyk** [comment]
Please test this branch thoroughly using ANDROID_LOCATION_TEST.md. Pay attention to `adb logcat --pid="$(adb shell pidof -s app.organicmaps.debug)"|grep onMyPositionModeChanged` messages.

File: android/app/src/main/java/app/organicmaps/location/LocationState.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rtsisyk) [reply android/app/src/main/java/app/organicmaps/location/LocationState.java:73]: After #6465 NO_FOLLOW_NO_LOCATION became closer to LOCATION_IS_DISABLED, but there is still a case when drape changes from NO_FOLLOW_NO_LOCATION to PENDING_LOCATION after ~30h. This patch doesn't address this issue, but makes Android to be aware of any random changes of location mode that happen on the Drape side.

**Review by biodranik — APPROVED** [review]
Let's test it thoroughly in the master, @Jean-BaptisteC I assume that there could be some issues in edge cases with location (including API calls). We had a location test plan somewhere in docs, right?

**Comment by rtsisyk** [comment]
> We had a location test plan somewhere in docs, right?

https://github.com/organicmaps/organicmaps/blob/master/docs/ANDROID_LOCATION_TEST.md

**Review by Jean-BaptisteC — COMMENTED** [review]
✅ Pixel 6 - Android 14
